src/gre.rs

```diff
@@ -0,0 +1,289 @@
+/// is determined by the `checksum_present` flag. The `header_len()` method can be
+/// used to determine the actual length of the header at runtime (4 or 8 bytes).
+///
+/// For more details, see RFC 2784: https://www.rfc-editor.org/rfc/rfc2784.html
+/// 
+/// /// A struct containing the optional checksum and reserved fields.
+
+#[repr(C, packed)]
+#[derive(Debug, Copy, Clone)]
+pub struct GreOptionalFields {
+    /// This field is only valid if the Checksum Present flag is set.
+    pub check: [u8; 2],
+    /// A reserved field for future use, which MUST be transmitted as zero (optional).
+    /// This field is only present if the Checksum Present flag is set.
+    pub _reserved1: [u8; 2],
+}
+
+/// A union representing the 4 bytes that can either be optional GRE fields
+/// or the start of the payload data.
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub union GreDataUnion {
+    /// The interpretation when checksum_present() is true.
+    fields: GreOptionalFields,
+    /// The interpretation when checksum_present() is false.
+    payload_start: [u8; 4],
+}
+#[repr(C, packed)]
+#[derive(Copy, Clone)]
+#[cfg_attr(feature = "serde", derive(::serde::Serialize, ::serde::Deserialize))]
+pub struct GreHdr {
+    /// A 16-bit field containing the Checksum Present flag (1 bit),
+    /// Reserved0 (12 bits), and Version (3 bits).
+    /// In a compliant packet, Reserved0 and Version MUST be 0.
+    pub flag_reserved0_ver: [u8; 2],
+    /// The protocol type of the encapsulated payload packet.
+    pub protocol_type: EtherType,
+    /// This field is only valid if the Checksum Present flag is set.
+    pub data: GreDataUnion,
+}
+
+impl GreHdr {
+
+    pub const LEN: usize = size_of::<GreHdr>();
+    
+    /// Checks if the Checksum Present bit (C) is set.
+    #[inline]
+    pub fn checksum_present(&self) -> bool {
+        (self.flag_reserved0_ver[0] & 0x80) != 0
+    }
+
+    /// Sets or clears the Checksum Present bit (C).
+    #[inline]
+    pub fn set_checksum_present(&mut self, present: bool) {
+        if present {
+            self.flag_reserved0_ver[0] |= 0x80;
+        } else {
+            self.flag_reserved0_ver[0] &= !0x80;
+        }
+    }
+
+    /// Gets the 12-bit Reserved0 field.
+    #[inline]
+    pub fn get_reserved0(&self) -> u16 {
+        let upper = ((self.flag_reserved0_ver[1] & 0xF8) >> 3) as u16;
+        let lower = (self.flag_reserved0_ver[0] & 0x7F) as u16;
+        
+        (upper << 7) | lower
+    }
+
+    /// Sets the 12-bit Reserved0 field.
+    /// According to RFC 2784, this value SHOULD be set to 0.
+    #[inline]
+    pub fn set_reserved0(&mut self, value: u16) {
+        let value = value & 0x0FFF;
+
+        let upper = (value >> 7) as u8;
+        let lower = (value & 0x007F) as u8;
+        
+        self.flag_reserved0_ver[0] = (self.flag_reserved0_ver[0] & 0x80) | lower;
+        self.flag_reserved0_ver[1] = (self.flag_reserved0_ver[1] & 0x07) | (upper << 3);
+    }
```

Comment:
It seems appropriate to remove these functions because these are reserved fields with no real value within an ebpf context.